feat/40: declare product compatibility

Sidsector9 opened the pull request:
### All Submissions:

<!-- Mark completed items with an [x] -->
* [ ] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #40 .
Depends on https://github.com/woocommerce/woocommerce/issues/45143

### Steps to test the changes in this Pull Request:
<!-- Describe the steps to replicate the issue and confirm the fix -->
<!-- Try to include as many details as possible. -->


### Changelog entry
<!-- 
Each line should start with change type prefix`(Add|Fix|Dev) - `.
If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.
Add the `changelog: none` label if no changelog entry is needed.
-->

> Add - Add support for the WooCommerce Product Block Editor.

Sidsector9 committed:
add sync with square field

Sidsector9 committed:
replace core with custom blocks

Sidsector9 committed:
product editor compatibility

Sidsector9 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-square into feat/40

Sidsector9 committed:
add custom product compatibility fields

Sidsector9 committed:
update method name

jeffpaul commented:
Noting that this work is blocked by https://github.com/woocommerce/woocommerce/issues/45143

Sidsector9 committed:
Merge branch 'trunk' into feat/40

jeffpaul commented:
@Sidsector9 is this unblocked now that https://github.com/woocommerce/woocommerce/pull/46741 is merged (and part of Woo 8.9.0)?

Sidsector9 commented:
@jeffpaul I'll go through the upstream thread and confirm if it unblocks.

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
Merge branch 'feat/40' of github.com:woocommerce/woocommerce-square into feat/40

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
add minor changes for field visibility

Sidsector9 committed:
add phpcs fixes

Sidsector9 committed:
add product editor compat for variations

dkotter commented on includes/Admin/Product_Editor_Compatibility.php:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	/**
+	 * Adds pre-orders meta to the product response.
```
I don't think this is an accurate description of what this is doing

dkotter commented on includes/Admin/Product_Editor_Compatibility.php:
```diff
@@ -0,0 +1,181 @@
+		);
+
+		if ( in_array( $block->get_id(), $blocks_to_remove, true ) ) {
+			$block->remove(); // This is a workaround and is not recommended by Woo.
```
Do we have more information on why we're taking this approach if it's not recommended?

Sidsector9 replied:
@dkotter in the classic editor, the plugin modifies the functionality of some of the core product meta fields. In the new product editor, we don't have a way to modify them however we can remove them and implement our own custom implementation.

For example, in the classic editor we conditionally disable the inventory tracking field which at the moment we cannot do in the new editor.

dkotter replied:
I think it would be worth updating the comment here to note this, rather than just saying this isn't recommended, just so there's no confusion by anyone else that views this

dkotter commented on package.json:
```diff
@@ -69,7 +69,9 @@
   "dependencies": {
-    "@wordpress/data": "^8.1.0",
+    "@woocommerce/product-editor": "^1.4.0",
+    "@wordpress/data": "^7.6.0",
```
Looks like we downgraded this dependency. Curious if there's a reason why we needed to do that?

Sidsector9 replied:
There's a React peer dependency conflict after upgrading wordpress/scripts. This is why wordpress/data has been downgraded.

I'll fix all of that in #110 by using [Woo Dependency Extraction](https://www.npmjs.com/package/@woocommerce/dependency-extraction-webpack-plugin).

dkotter reviewed (changes requested):
Few minor things I've flagged.

Couple other questions:

1. Are there any new E2E tests we want to / should add?
2. If I'm following the dependencies here correctly, seems the work here relies on Woo 8.9, is that correct? If so, do we need to wait to release this until 8.9 becomes our minimum?

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
Merge branch 'feat/40' of github.com:woocommerce/woocommerce-square into feat/40

Sidsector9 committed:
improve documentation

Sidsector9 committed:
add test for product editor

Sidsector9 committed:
add test to verify inventory sync

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
Merge branch 'feat/40' of github.com:woocommerce/woocommerce-square into feat/40

Sidsector9 commented:
@dkotter responding to your questions:

> Are there any new E2E tests we want to / should add?

I've added the tests.

> If I'm following the dependencies here correctly, seems the work here relies on Woo 8.9, is that correct? If so, do we need to wait to release this until 8.9 becomes our minimum?

That's correct!

Sidsector9 committed:
Merge branch 'trunk' into feat/40

qasumitbagthariya commented:
### QA Update ❌
----

@Sidsector9 
I'm getting an issue with the checkout block. Can you please check from your end? 

![image](https://github.com/woocommerce/woocommerce-square/assets/67687255/1fe49755-33c6-4647-90f0-1401e8575ce4)

Let me know if you need any help to reproduce the issue.

Sidsector9 committed:
Merge branch 'feat/40' of github.com:woocommerce/woocommerce-square into feat/40

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
fix broken checkout block

qasumitbagthariya commented:
**QA Update ❌**
---
@Sidsector9 

**Issue:1**
When the "Sync inventory" setting is disabled, the stock levels for products should not synchronize with Square. However, in the new product editor, the sync setting still appears, which is inconsistent with the behavior observed in the old product editor where this setting is not displayed.

**How to reproduce:**
1. Disable the "Sync inventory" setting in the square settings.
2. SOR - Square or Woocommerce
3. Navigate and create a new product with the new product editor.
4. Observe the presence of the inventory sync setting.
5. Compare this behavior with the old product editor where the inventory sync setting is correctly hidden.


https://github.com/woocommerce/woocommerce-square/assets/67687255/c2505672-1256-42cd-986b-e657a6d8dc98

---

**Issue:2** 

In the old product editor, there are two options for synchronization: "Synced with Square" and "Sync inventory". However, in the new product editor, only the "Synced with Square" option is available. whether this change is intentional or is there is a missing logic behind the sync options. Please let me know if anything I missed here.

SOR - Square or Woocommerce

![image](https://github.com/woocommerce/woocommerce-square/assets/67687255/d99ddd6c-d570-4c05-9cb6-e987553b0b3b)

![image](https://github.com/woocommerce/woocommerce-square/assets/67687255/da28d9f8-98ca-42ad-adb1-773f0653464b)

Sidsector9 committed:
hide inventory fields

When inventory tracking is disabled and a product is synced in Woo,
we hide the field once the syncing process is complete.

Sidsector9 committed:
update conditions for stock management field

Sidsector9 committed:
remove redundant logic

Sidsector9 committed:
hiden sync features when product isn't published

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
Merge branch 'trunk' into feat/40

qasumitbagthariya commented:
@Sidsector9 I'm facing a fatal error in fix branch `feat/40` can you please check from your end.

![image](https://github.com/woocommerce/woocommerce-square/assets/67687255/310d9080-266a-4863-ad7f-618f9a5873fa)

Sidsector9 committed:
fix callback

qasumitbagthariya commented:
### QA Update ❌
-------------

@Sidsector9 Currently, there is no checkbox visible for gift card products, So need to add support for gift card products.

**New product editor**

![image](https://github.com/woocommerce/woocommerce-square/assets/67687255/210e85e7-7fd2-4f6e-9c6a-1c9226ad9d78)


![image](https://github.com/woocommerce/woocommerce-square/assets/67687255/53b38efe-dc37-4838-8df6-e159918c2c6c)

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
add gift card support

Sidsector9 committed:
check if sor is disabled

Sidsector9 committed:
add sync inventory button

Sidsector9 committed:
conditionally add gift card setting

vikrampm1 commented:
@sid is this being worked on or ready for another round of QA?

Sidsector9 commented:
@vikrampm1 this is in QA.

Sidsector9 committed:
Merge branch 'trunk' into feat/40

qasumitbagthariya commented:
### QA Update ✅ 

-------------

I have checked this issue in the `feat/40` branch and it is working as expected.

I tested the following on this branch:

- New product editor 

**Testing Environment** 

<Details>

- WordPress: 6.6
- Theme: Storefront 4.6.0
- Theme: Twenty Twenty-Four 1.2
- WooCommerce - 9.1.2
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 

</Details>


https://github.com/user-attachments/assets/bcf578ab-61f3-40a8-b0cd-0459a161291c


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [ ] Yes
- [x] Not Required/Applicable for this PR

Sidsector9 committed:
Merge branch 'feat/40' of github.com:woocommerce/woocommerce-square into feat/40

Sidsector9 committed:
Merge branch 'trunk' into feat/40

wjrosa commented on includes/Admin/Product_Editor_Compatibility.php:
```diff
@@ -0,0 +1,241 @@
+		 */
+		$parent = $sku_field->get_parent();
+
+		$is_inventiry_sync_enabled = wc_square()->get_settings_handler()->is_inventory_sync_enabled();
```
```suggestion
		$is_inventory_sync_enabled = wc_square()->get_settings_handler()->is_inventory_sync_enabled();
```

wjrosa commented on includes/Admin/Product_Editor_Compatibility.php:
```diff
@@ -0,0 +1,241 @@
+
+		$is_inventiry_sync_enabled = wc_square()->get_settings_handler()->is_inventory_sync_enabled();
+
+		if ( ! $is_inventiry_sync_enabled ) {
```
```suggestion
		if ( ! $is_inventory_sync_enabled ) {
```

wjrosa reviewed (approved):
Code looks good. Left some minor suggestions 👍

vikrampm1 commented:
@Sidsector9 some minor suggestions to resolve and then we can move this to UAT.

Sidsector9 committed:
fix minor typo

Sidsector9 committed:
fix eslint errors

jeffpaul commented on package.json:
```diff
@@ -37,7 +37,7 @@
     "@woocommerce/eslint-plugin": "^2.2.0",
     "@wordpress/env": "^10.0.0",
-    "@wordpress/scripts": "^27.9.0",
+    "@wordpress/scripts": "^24.6.0",
```
This downgrade is purposeful I assume?

Sidsector9 replied:
@jeffpaul yes, React version mismatch

jeffpaul commented on src/admin/product-blocks/sync-with-square-field/edit.tsx:
```diff
@@ -0,0 +1,94 @@
+		if ( 'simple' === product_type && ! sku.length ) {
+			setSkuError(
+				sprintf(
+					__( `Please add an SKU to sync %s with Square. The SKU must match the item's SKU in your Square account.`, 'woocommerce-square' ),
```
Please update to "Please add a SKU..."

Sidsector9 replied:
@jeffpaul although SKU starts with a consonant, I think it will still use an "an" as "SKU" is typically pronounced as "ess-kay-you," starting with a vowel sound ("ess").

jeffpaul commented on src/admin/product-blocks/stock-quantity-field/edit.js:
```diff
@@ -0,0 +1,228 @@
+/**
+ * External dependencies
+ */
+import { useWooBlockProps } from '@woocommerce/block-templates';
```
Looks like eslint warning that we should be able to resolve here

Sidsector9 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-square into trunk

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
Merge branch 'trunk' into feat/40

Sidsector9 committed:
fix eslint dependency error

Sidsector9 committed:
eslint error fixes

qasumitbagthariya commented:
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.5.5, npm version 8.19.4, node version 16.20.0)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

**Testing Environment** 

<Details>

- WordPress: 6.6.1
- Theme: Storefront 4.6.0
- Theme: Twenty Twenty-Four 1.2
- WooCommerce - 9.1.4
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Branch: smoke-testing

</Details>



Next Step- Ready to Merge 🚀

qasumitbagthariya commented:
@Sidsector9 

I encountered a sync failure while regression in the following scenario where the SOR is changed from Square to WooCommerce.

**Steps to Reproduce:**

1. Set the SOR to Square
2. Create a variable product in the Square Dashboard.
3. import the product.
4. After the variable product is imported successfully, change the SOR from Square to WooCommerce.
5. Run a manual sync.

<img width="862" alt="image" src="https://github.com/user-attachments/assets/49eecd0d-0d25-4988-b237-c027fa1d1f75">

**Testing Environment** 

<Details>

- WordPress: 6.6.1
- Theme: Twenty Twenty-Four 1.2
- WooCommerce - 9.1.4
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Branch: smoke-testing

</Details>

qasumitbagthariya commented:
@Sidsector9 Approving the regression testing, as I have created a separate [issue](https://github.com/woocommerce/woocommerce-square/issues/199) 👍

Sidsector9 committed:
Merge branch 'trunk' into feat/40

dkotter committed:
Merge branch 'trunk' into feat/40
